File: src/Scripting/Ruby/chain_of_responsibility.rb
class Handler
  def initialize(name, limit)
    @name = name
    @limit = limit
    @next_handler = nil
  end

  def then_next(handler)
    @next_handler = handler
    handler
  end

  def handle(amount, visited)
    visited << @name
    return @name if amount <= @limit || @next_handler.nil?

    @next_handler.handle(amount, visited)
  end
end

faq = Handler.new('faq', 50)
billing = Handler.new('billing', 500)
escalation = Handler.new('escalation', Float::INFINITY)
faq.then_next(billing).then_next(escalation)

visited = []
handled = faq.handle(250, visited)
puts "visited=#{visited.join('>')};handled=#{handled};result=refund(250)"

File: src/Scripting/Perl/chain_of_responsibility.pl
use strict;
use warnings;

my @visited;
my @handlers = (
    { name => 'faq',        limit => 50 },
    { name => 'billing',    limit => 500 },
    { name => 'escalation', limit => undef },
);

sub handle_refund {
    my ($amount) = @_;
    for my $handler (@handlers) {
        push @visited, $handler->{name};
        if (!defined $handler->{limit} || $amount <= $handler->{limit}) {
            return ($handler->{name}, "refund($amount)");
        }
    }
    die "unhandled request\n";
}

my ($handled, $result) = handle_refund(250);
print 'visited=' . join('>', @visited) . ";handled=$handled;result=$result\n";
